crypto/gmac/gmac.c

```diff
@@ -0,0 +1,168 @@
+}
+
+static int gmac_update(EVP_MAC_IMPL *gctx, const unsigned char *data,
+                       size_t datalen)
+{
+    EVP_CIPHER_CTX *ctx = gctx->ctx;
+    int outlen;
+
+    while (datalen > INT_MAX) {
+        if (!EVP_EncryptUpdate(ctx, NULL, &outlen, data, INT_MAX))
+            return 0;
+        data += INT_MAX;
+        datalen -= INT_MAX;
+    }
+    return EVP_EncryptUpdate(ctx, NULL, &outlen, data, datalen);
+}
+
+static int gmac_final(EVP_MAC_IMPL *gctx, unsigned char *out)
+{
+    int hlen;
+
+    if (!EVP_EncryptFinal_ex(gctx->ctx, out, &hlen)
+        || !EVP_CIPHER_CTX_ctrl(gctx->ctx, EVP_CTRL_AEAD_GET_TAG,
+                                gmac_size(gctx), out))
+        return 0;
+    return 1;
+}
+
+static int gmac_ctrl(EVP_MAC_IMPL *gctx, int cmd, va_list args)
+{
+    const unsigned char *p;
+    size_t len;
+    EVP_CIPHER_CTX *ctx = gctx->ctx;
+    const EVP_CIPHER *cipher;
+    ENGINE *engine;
+
+    switch (cmd) {
+    case EVP_MAC_CTRL_SET_CIPHER:
+        cipher = va_arg(args, const EVP_CIPHER *);
+        return (EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER)
+                && EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL);
+
+    case EVP_MAC_CTRL_SET_KEY:
+        p = va_arg(args, const unsigned char *);
+        len = va_arg(args, size_t);
+        return len == (size_t)EVP_CIPHER_CTX_key_length(ctx)
+               && EVP_EncryptInit_ex(ctx, NULL, NULL, p, NULL);
+
+    case EVP_MAC_CTRL_SET_IV:
+        p = va_arg(args, const unsigned char *);
+        len = va_arg(args, size_t);
+        return EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, len, NULL)
+               && EVP_EncryptInit_ex(ctx, NULL, NULL, NULL, p);
+
+    case EVP_MAC_CTRL_SET_ENGINE:
+        engine = va_arg(args, ENGINE *);
+        return EVP_EncryptInit_ex(ctx, NULL, engine, NULL, NULL);
+
+    default:
+        return -2;
+    }
+}
+
+static int gmac_ctrl_int(EVP_MAC_IMPL *gctx, int cmd, ...)
+{
+    int rv;
+    va_list args;
+
+    va_start(args, cmd);
+    rv = gmac_ctrl(gctx, cmd, args);
+    va_end(args);
+
+    return rv;
+}
+
+static int gmac_ctrl_str_cb(void *gctx, int cmd, void *buf, size_t buflen)
+{
+    return gmac_ctrl_int(gctx, cmd, buf, buflen);
+}
+
+static int gmac_ctrl_str(EVP_MAC_IMPL *gctx, const char *type,
+                         const char *value)
+{
+    if (!value)
+        return 0;
+    if (strcmp(type, "key") == 0)
+        return EVP_str2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_KEY,
+                            value);
+    else if (strcmp(type, "hexkey") == 0)
+        return EVP_hex2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_KEY,
+                            value);
+    else if (strcmp(type, "iv") == 0)
+        return EVP_str2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
+    else if (strcmp(type, "hexiv") == 0)
+        return EVP_hex2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
+    return -2;
```

Comment:
It should be possible to set the cipher as well, no?  CMAC has that possibility...

Reply:
Oops, I missed that bit.  I do allow the cipher via the vararg call.  I'll add.